### PR TITLE
Feature flux controlnet img2img and inpaint pipeline

### LINKED ISSUE #9402: [Flux ControlNet] Add img2img and inpaint pipelines
We recently added img2img and inpainting pipelines for Flux thanks to @Gothos contribution. 

We also have controlnet support for Flux thanks to @wangqixun.

It'd be nice to have controlnet versions of these pipelines since there's been requests to have them.

Basically, we need to create two new pipelines that add the controlnet support from this [pipeline ](https://github.com/huggingface/diffusers/blob/f28a8c257afe8eeb16b4deb973c6b1829f6aea59/src/diffusers/pipelines/flux/pipeline_flux_controlnet.py) to the corresponding pipellines.

- [ ] [Image to image](https://github.com/huggingface/diffusers/blob/f28a8c257afe8eeb16b4deb973c6b1829f6aea59/src/diffusers/pipelines/flux/pipeline_flux_img2img.py)
- [ ] [Inpaint](https://github.com/huggingface/diffusers/blob/f28a8c257afe8eeb16b4deb973c6b1829f6aea59/src/diffusers/pipelines/flux/pipeline_flux_inpaint.py)

Related issue: #9158 

Let me know if someone is interested in contributing this.

### PR DESCRIPTION
**Pull request by ighoshsubho**
# What does this PR do?

Fixes #9402 and #9158

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by ighoshsubho** [comment]
Hey @a-r-r-o-w can you suggest if the flux img2img controlnet pipeline needs any fix or if this architecture is fine to go with.

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet_image_to_image.py [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+            )
+        scheduler.set_timesteps(sigmas=sigmas, device=device, **kwargs)
+        timesteps = scheduler.timesteps
+        num_inference_steps = len(timesteps)
+    else:
+        scheduler.set_timesteps(num_inference_steps, device=device, **kwargs)
+        timesteps = scheduler.timesteps
+    return timesteps, num_inference_steps
+
+class FluxControlNetImg2ImgPipeline(FluxControlNetPipeline):
```
[the remark this answers is not in the document]
Comment (ighoshsubho) [reply src/diffusers/pipelines/flux/pipeline_flux_controlnet_image_to_image.py:83]: sure will make an update for this

**Review by yiyixuxu — APPROVED** [review]
thanks!

**Comment by yiyixuxu** [comment]
awesome! thank you!
can you share example script + result from these two pipelines?

**Comment by yiyixuxu** [comment]
can you run `make style ` and `fix copies` so CI can pass?

**Comment by ighoshsubho** [comment]
> awesome! thank you!
> can you share example script + result from these two pipelines?

Sure I will!

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet_image_to_image.py [hunk 2]
```diff
@@ -0,0 +1,882 @@
...
+        width = width // vae_scale_factor
+
+        latents = latents.view(batch_size, height, width, channels // 4, 2, 2)
+        latents = latents.permute(0, 3, 1, 4, 2, 5)
+
+        latents = latents.reshape(batch_size, channels // (2 * 2), height * 2, width * 2)
+
+        return latents
+    
+    def prepare_latents(
```
[the remark this answers is not in the document]
Comment (ighoshsubho) [reply src/diffusers/pipelines/flux/pipeline_flux_controlnet_image_to_image.py:521]: will do

**Comment by ighoshsubho** [comment]
> awesome! thank you! can you share example script + result from these two pipelines?



> awesome! thank you! can you share example script + result from these two pipelines?

Here is example for img2img - https://github.com/huggingface/diffusers/blob/a30ca655dc597aacc7f9e016458bae4355eb7a62/src/diffusers/pipelines/flux/pipeline_flux_controlnet_image_to_image.py#L42

Here is example for inpaint - https://github.com/huggingface/diffusers/blob/a30ca655dc597aacc7f9e016458bae4355eb7a62/src/diffusers/pipelines/flux/pipeline_flux_controlnet_inpainting.py#L43

**Comment by ighoshsubho** [comment]
@yiyixuxu any updates, are we good to go or does it require further changes

**Comment by yiyixuxu** [comment]
can we look into the failing fast tests?

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet_inpainting.py [hunk 3]
```diff
@@ -0,0 +1,1112 @@
...
+        scheduler.set_timesteps(sigmas=sigmas, device=device, **kwargs)
+        timesteps = scheduler.timesteps
+        num_inference_steps = len(timesteps)
+    else:
+        scheduler.set_timesteps(num_inference_steps, device=device, **kwargs)
+        timesteps = scheduler.timesteps
+    return timesteps, num_inference_steps
+
+
+class FluxControlNetInpaintPipeline(DiffusionPipeline, FluxLoraLoaderMixin, FromSingleFileMixin):
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/flux/pipeline_flux_controlnet_inpainting.py:178]: Could you add the docstring here similar to [this](https://github.com/huggingface/diffusers/blob/48e36353d8cbf0322ec1ad0684b95d11f70af2de/src/diffusers/pipelines/flux/pipeline_flux.py#L142)?

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet_image_to_image.py [hunk 4]
```diff
@@ -0,0 +1,923 @@
...
+        scheduler.set_timesteps(sigmas=sigmas, device=device, **kwargs)
+        timesteps = scheduler.timesteps
+        num_inference_steps = len(timesteps)
+    else:
+        scheduler.set_timesteps(num_inference_steps, device=device, **kwargs)
+        timesteps = scheduler.timesteps
+    return timesteps, num_inference_steps
+
+
+class FluxControlNetImg2ImgPipeline(DiffusionPipeline, FluxLoraLoaderMixin, FromSingleFileMixin):
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/flux/pipeline_flux_controlnet_image_to_image.py:170]: Docstring here too

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet_image_to_image.py [hunk 5]
```diff
@@ -0,0 +1,923 @@
...
+                The list of tensor inputs for the `callback_on_step_end` function.
+            max_sequence_length (`int`, *optional*, defaults to 512):
+                The maximum length of the sequence to be generated.
+
+        Returns:
+            [`~pipelines.flux.FluxPipelineOutput`] or `tuple`: [`~pipelines.flux.FluxPipelineOutput`] if `return_dict`
+            is True, otherwise a `tuple`. When returning a tuple, the first element is a list with the generated
+            images.
+
+        Examples:
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/flux/pipeline_flux_controlnet_image_to_image.py:682]: Usually, we have the Examples before the Returns. Let us know if you saw a difference somewhere

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet_inpainting.py [hunk 6]
```diff
@@ -0,0 +1,1112 @@
...
+                The list of tensor inputs for the `callback_on_step_end` function.
+            max_sequence_length (`int`, *optional*, defaults to 512):
+                The maximum length of the sequence to be generated.
+
+        Returns:
+            [`~pipelines.flux.FluxPipelineOutput`] or `tuple`: [`~pipelines.flux.FluxPipelineOutput`] if `return_dict`
+            is True, otherwise a `tuple`. When returning a tuple, the first element is a list with the generated
+            images.
+
+        Examples:
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/flux/pipeline_flux_controlnet_inpainting.py:796]: Same comment on Examples being before Returns

**Review by a-r-r-o-w — COMMENTED** [review]
Thank you, the PR looks very clean and everything seems right! Great work @ighoshsubho 

Just some minor nits from me. Could you also add these pipelines to the flux [docs](https://github.com/huggingface/diffusers/blob/main/docs/source/en/api/pipelines/flux.md) and address the failing tests?

Btw @yiyixuxu, do we need the slow integration tests here? I think we don't have them on derived pipelines that are made from two or more components (such as Flux + ImageToImage, or AnimateDiff + VideoToVideo)

**Comment by ighoshsubho** [comment]
> Thank you, the PR looks very clean and everything seems right! Great work @ighoshsubho
> 
> Just some minor nits from me. Could you also add these pipelines to the flux [docs](https://github.com/huggingface/diffusers/blob/main/docs/source/en/api/pipelines/flux.md) and address the failing tests?
> 
> Btw @yiyixuxu, do we need the slow integration tests here? I think we don't have them on derived pipelines that are made from two or more components (such as Flux + ImageToImage, or AnimateDiff + VideoToVideo)

will address the tests and minor fixes

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet_inpainting.py (hunk 3 above)
Comment (ighoshsubho) [reply src/diffusers/pipelines/flux/pipeline_flux_controlnet_inpainting.py:178]: Fixed now

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet_image_to_image.py (hunk 4 above)
Comment (ighoshsubho) [reply src/diffusers/pipelines/flux/pipeline_flux_controlnet_image_to_image.py:170]: Fixed now

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet_image_to_image.py (hunk 5 above)
Comment (ighoshsubho) [reply src/diffusers/pipelines/flux/pipeline_flux_controlnet_image_to_image.py]: Fixed

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet_inpainting.py (hunk 6 above)
Comment (ighoshsubho) [reply src/diffusers/pipelines/flux/pipeline_flux_controlnet_inpainting.py]: This is fixed too

File: docs/source/en/api/pipelines/flux.md [hunk 7]
```diff
@@ -160,18 +162,115 @@ image.save("flux-fp8-dev.png")
 
 ## FluxPipeline
 
-[[autodoc]] FluxPipeline
-	- all
-	- __call__
+[[autodoc]] FluxPipeline - all - **call**
```
Comment (a-r-r-o-w) [inline docs/source/en/api/pipelines/flux.md:165]: Not sure why these were changed, please revert to keep the scope of this PR limited

File: docs/source/en/api/pipelines/flux.md [hunk 8]
```diff
@@ -160,18 +162,115 @@ image.save("flux-fp8-dev.png")
...
+[[autodoc]] FluxImg2ImgPipeline - all - **call**
 
 ## FluxInpaintPipeline
 
-[[autodoc]] FluxInpaintPipeline
-	- all
-	- __call__
+[[autodoc]] FluxInpaintPipeline - all - **call**
+
+## Flux ControlNet Inpaint Pipeline
```
Comment (a-r-r-o-w) [inline docs/source/en/api/pipelines/flux.md:175]: We don't need to provide an example here. autodoc will take the example from the `EXAMPLE_DOC_STRING` defined at the top of the pipeline file here and show the relevant usage.

**Review by a-r-r-o-w — COMMENTED** [review]
Just two more change requests

File: docs/source/en/api/pipelines/flux.md (hunk 7 above)
Comment (ighoshsubho) [reply docs/source/en/api/pipelines/flux.md:165]: sure will do

File: docs/source/en/api/pipelines/flux.md (hunk 8 above)
Comment (ighoshsubho) [reply docs/source/en/api/pipelines/flux.md:175]: sure

**Comment by ighoshsubho** [comment]
@a-r-r-o-w fixed the minor bugs and tests should pass now

**Comment by ighoshsubho** [comment]
@yiyixuxu and @a-r-r-o-w fixed the minor bugs and tests should pass now

**Comment by mrtpk** [comment]
Can't wait for this awesomeness!!!!

**Comment by yiyixuxu** [comment]
thank you!

**Comment by mrtpk** [comment]
Thank you @ighoshsubho

**Comment by DoiryCool** [comment]
How can I use other controlnets, such as XLabs-AI/flux-controlnet-canny-v3? The canny model in the example exceeds my vmem

**Pull request merged**